### PR TITLE
Add support for graph.RO_QUERY command

### PR DESCRIPTION
**Pull request by filipecosta90**
This PR kicks off the support for graph.RO_QUERY command by introducing queryReadOnly(). It includes the same api methods that query() offers.

### CODE REVIEW

File: src/main/java/com/redislabs/redisgraph/RedisGraph.java [hunk 1]
```diff
@@ -34,6 +52,17 @@
...
+     * Execute a Cypher read-only query with arguments
+     * @param graphId a graph to perform the query on
+     * @param query Cypher query
+     * @param args
+     * @return a result set
+     * @deprecated use {@link #query(String, String, Map)} instead.
+     */
+    @Deprecated
+    ResultSet queryReadOnly(String graphId, String query, Object ...args);
+
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/RedisGraph.java:65]: remove this as this is deprecated API

File: src/main/java/com/redislabs/redisgraph/impl/api/AbstractRedisGraph.java [hunk 2]
```diff
@@ -72,6 +111,20 @@ public ResultSet query(String graphId, String query, Object ...args) {
...
+     * @param args
+     * @return a result set
+     * @deprecated use {@link #query(String, String, Map)} instead.
+     */
+    @Deprecated
+    public ResultSet queryReadOnly(String graphId, String query, Object ...args) {
+        String preparedQuery = Utils.prepareQuery(query, args);
+        return sendReadOnlyQuery(graphId, preparedQuery);
+    }
+
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/impl/api/AbstractRedisGraph.java:127]: deprecated

File: src/main/java/com/redislabs/redisgraph/RedisGraph.java [hunk 3]
```diff
@@ -14,6 +14,14 @@
      */
     ResultSet query(String graphId, String query);
 
+    /**
+     * Execute a Cypher read-only query.
+     * @param graphId a graph to perform the query on
+     * @param query Cypher query
+     * @return a result set
+     */
+    ResultSet queryReadOnly(String graphId, String query);
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/RedisGraph.java:23]:
```suggestion
    ResultSet readOnlyQuery(String graphId, String query);
```

File: src/main/java/com/redislabs/redisgraph/RedisGraph.java [hunk 4]
```diff
@@ -23,6 +31,16 @@
...
     ResultSet query(String graphId, String query, long timeout);
 
+    /**
+     * Execute a Cypher read-only query with timeout.
+     * @param graphId a graph to perform the query on
+     * @param query Cypher query
+     * @param timeout
+     * @return a result set
+     */
+    ResultSet queryReadOnly(String graphId, String query, long timeout);
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/RedisGraph.java:41]:
```suggestion
    ResultSet readOnlyQuery(String graphId, String query, long timeout);
```

File: src/main/java/com/redislabs/redisgraph/RedisGraph.java [hunk 5]
```diff
@@ -43,6 +72,15 @@
...
     ResultSet query(String graphId, String query, Map<String, Object> params);
 
+    /**
+     * Executes a cypher read-only query with parameters.
+     * @param graphId a graph to perform the query on.
+     * @param query Cypher query.
+     * @param params parameters map.
+     * @return a result set.
+     */
+    ResultSet queryReadOnly(String graphId, String query, Map<String, Object> params);
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/RedisGraph.java:82]:
```suggestion
    ResultSet readOnlyQuery(String graphId, String query, Map<String, Object> params);
```

File: src/main/java/com/redislabs/redisgraph/RedisGraph.java [hunk 6]
```diff
@@ -53,6 +91,16 @@
...
 
+    /**
+     * Executes a cypher read-only query with parameters and timeout.
+     * @param graphId a graph to perform the query on.
+     * @param query Cypher query.
+     * @param params parameters map.
+     * @param timeout
+     * @return a result set.
+     */
+    ResultSet queryReadOnly(String graphId, String query, Map<String, Object> params, long timeout);
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/RedisGraph.java:102]:
```suggestion
    ResultSet readOnlyQuery(String graphId, String query, Map<String, Object> params, long timeout);
```

**Review by DvirDukhan — CHANGES_REQUESTED** [review]
@filipecosta90 
1. see comments
2. please add the RO API also to the transactions 
3. I think that we need dedicated test functions for RO API instead of mixing with the current tests, which are messy

File: src/main/java/com/redislabs/redisgraph/impl/api/AbstractRedisGraph.java (hunk 2 above)
Comment (filipecosta90) [reply src/main/java/com/redislabs/redisgraph/impl/api/AbstractRedisGraph.java]: fixed in the latest commit

**Review by sazzad16 — APPROVED** [review]

**Comment by tanvirp** [comment]
Will RO api work with read only slaves? Looks like https://github.com/RedisGraph/JRedisGraph/blob/ed0c37c6708709ec0a27140897ac1a496b586010/src/main/java/com/redislabs/redisgraph/impl/graph_cache/GraphCacheList.java#L53 will eventually call https://github.com/RedisGraph/JRedisGraph/blob/ed0c37c6708709ec0a27140897ac1a496b586010/src/main/java/com/redislabs/redisgraph/impl/api/AbstractRedisGraph.java#L112, which will issue the `QUERY` command instead of `RO_QUERY` command. I am not sure if this will be taken care in this PR.

**Review by sazzad16 — APPROVED** [review]

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java [hunk 7]
```diff
@@ -59,6 +59,27 @@ protected ResultSet sendQuery(String graphId, String preparedQuery) {
...
+        try {
+            List<Object> rawResponse = (List<Object>) conn.sendCommand(RedisGraphCommand.RO_QUERY, graphId, preparedQuery, Utils.COMPACT_STRING);
+            return new ResultSetImpl(rawResponse, this, caches.getGraphCache(graphId));
+        }
+        catch (JRedisGraphRunTimeException rt) {
+            throw rt;
+        }
+        catch (JedisDataException j) {
+            throw new JRedisGraphCompileTimeException(j);
+        }
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java:80]: Two of these error classes were removed. Now there is only `JRedisGraphException`.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Review by sazzad16 — APPROVED** [review]

**Comment by sazzad16** [comment]
@tanvirp Do you have a solution in mind? Feel free to craft a PR.

**Review by gkorland — APPROVED** [review]

**Pull request merged**